### ISSUE TITLE
Skorch with microsoft nni: need guide

### ISSUE THREAD

**Issue by UTUnex** [skorch-dev/skorch#685]
Hi, I have used skorch for some days and I really like this wrapper which makes DL model building & training easier. Recently I have been trying to use skorch with microsoft nni for hyperparameter tuning and process monitoring and I got stuck.
I want to use the BOHB advisor in NNI but I just don't know how to make it in skorch.  
https://github.com/microsoft/nni/tree/master/examples/trials/mnist-advisor
This is an example that NNI provided but it's implemented in tensorflow .  
Can someone here rewrite it in a skorch way? 
Thank you.

**Comment by BenjaminBossan** [comment]
I've never used NNI so I can only try to make educated guesses here.

First of all, one of the main advantages of using skorch is that it provides an sklearn API, meaning that you can use sklearn's `GridSearchCV`, `RandomizedSearchCV` and all other hyper-parameter search strategies that are compatible with sklearn's API. Therefore, I would look at how NNI works with sklearn and try to adjust from there. [Here](https://github.com/microsoft/nni/blob/master/docs/en_US/TrialExample/SklearnExamples.md) is an example in the NNI docs, but I don't know if it's possible to use your own sklearn estimator instead of only using the builtin ones (assume it is, otherwise the feature would be quite useless).

If you can provide some code and show the error message, that could also help in debugging.

**Comment by UTUnex** [comment]
Thank you for your reply. The code in the compression file is what I have implemented and it works but it's tuned by TPE.

[skorch_TPE.zip](https://github.com/skorch-dev/skorch/files/5146714/skorch_TPE.zip)

**Comment by UTUnex** [comment]
To adjust it to the hyperband or BOHB version, I think, first we need set the trial-budget, then we need report the intermediate result to nni; I'm confused what the intermediate result is if we set the budget as the number of  epochs. And ofc how to implement it. Could you please try to adjust it based on my code ?

**Comment by BenjaminBossan** [comment]
Hey, so I got your code to run by replacing the data in `load_data` with some dummy data, since I don't have your data. However, nothing really happens when I run the code (using `python main.py`) because `nni.get_next_parameter()` returns an empty dict for me. Why that is I don't know, since I never worked with nni.

Looking at your code, however, there are some things that could definitely be improved. E.g. you have a separate step to fit a `StandardScaler`, when really you should put the whole model into an sklearn `Pipeline`. Then you have a `get_model(PARAMS)` function which does things like this:

`model.module__dropout = PARAMS.get("dropout")`

However, that is unnecessary with skorch, you can just do

```python
net = NeuralNetClassifier(module=AntiFraudNet1)
net.set_params(**PARAMS)
```

For that to work, your `PARAMS` dict would need the right key names, though, e.g. `module__dropout` instead of `dropout`. If you combine this with a 'Pipeline`, you need to additionally prepend the name of the pipeline step, e.g.:

```python
from sklearn.pipeline import Pipeline

preprocessing = StandardScaler()
net = NeuralNetClassifier(module=AntiFraudNet1, verbose=0, train_split=False)
model = Pipeline([
    ('preproc', preprocessing),
    ('clf', net),  # the 'clf' string must be prepended
])

# later
params = {'clf__module__hidden_units': 100, 'clf__module__dropout': 0.5, 'lr': 0.01}
model.set_params(**params)
```

Maybe it would make sense for you to familiarize yourself a bit more with skorch (e.g. through [this notebook](https://nbviewer.jupyter.org/github/skorch-dev/skorch/blob/master/notebooks/Basic_Usage.ipynb)) and then re-visiting your code.

**Comment by UTUnex** [comment]
Sir, thank you very much for your advice. Yeah, I will recheck the code and try to refine it but now what I really hope is to make the nni work.
